addons/project/models/project.py

```diff
@@ -2696,10 +2696,54 @@ def search_read(self, domain=None, fields=None, offset=0, limit=None, order=None
 
     @api.model
     def _name_search(self, name='', args=None, operator='ilike', limit=100, name_get_uid=None):
-        domain = args
-        if 'project_id' in self.env.context:
-            domain = self._get_project_tags_domain(domain, self.env.context.get('project_id'))
-        return super()._name_search(name, domain, operator, limit, name_get_uid)
+        if 'project_id' in self.env.context and operator == 'ilike':
+            # `args` has the form of the default filter ['!', ['id', 'in', <ids>]]
+            # passed to exclude already selected tags -> exclude them in our query too
+            excluded_ids = list(args[1][2]) \
+                if args and len(args) == 2 and args[0] == '!' and len(args[1]) == 3 and args[1][:2] == ["id", "in"] \
+                else []
+            # UNION ALL is lazy evaluated, if the first query has enough results,
+            # the second is not executed (just planned).
+            query = """
+                WITH query_tags_in_tasks AS (
+                    SELECT tags.id, tags.name, 1 AS sequence
+                    FROM project_tags AS tags
+                    JOIN project_tags_project_task_rel AS tags__tasks_ids
+                        ON tags__tasks_ids.project_tags_id = tags.id
+                    JOIN project_task AS task
+                        ON task.id = tags__tasks_ids.project_task_id
+                        AND task.project_id = %(project_id)s
+                    WHERE tags.id != ALL(%(excluded_ids)s)
+                    AND COALESCE(tags.name ->> %(lang)s, tags.name ->> 'en_US') ILIKE %(search_term)s
+                    LIMIT %(limit)s
+                ), query_all_tags AS (
+                    SELECT tags.id, tags.name, 2 AS sequence
+                    FROM project_tags AS tags
+                    WHERE tags.id != ALL(%(excluded_ids)s)
+                    AND tags.id NOT IN (SELECT id FROM query_tags_in_tasks)
+                    AND COALESCE(tags.name ->> %(lang)s, tags.name ->> 'en_US') ILIKE %(search_term)s
+                    LIMIT %(limit)s
+                )
+                SELECT id FROM (
+                    SELECT id, name, sequence
+                    FROM query_tags_in_tasks
+                    UNION ALL
+                    SELECT id, name, sequence
+                    FROM query_all_tags
+                    LIMIT %(limit)s) AS tags
+                ORDER BY sequence, COALESCE(tags.name ->> %(lang)s, tags.name ->> 'en_US')
+            """
+            params = {
+                'project_id': self.env.context.get('project_id'),
+                'excluded_ids': excluded_ids,
+                'limit': limit,
+                'lang': self.env.context.get('lang', 'en_US'),
+                'search_term': '%' + name + '%',
```

Comment:
Finally I preferred to do the '%' formatting outside of SQL, because the alternative would have been to do
```sql
... ILIKE '%%' || %(search_term)s || '%%'
```
Because the `'%'` needs to be escaped, and the default escape in the `ILIKE` is `%`, leading to a `'%%'`, which may confuse future readers. Doing the concat python side is clear and unambiguous.